### PR TITLE
Support named sources for DynamicTest.stream(…)

### LINKED ISSUE #2375: Support named sources for DynamicTest.stream(…)
I am trying to create a `Stream<DynamicTest>` from a set of different sources (read: different types) and the way that generics work makes this quite cumbersome. Here's an example from Spring HATEOAS:

```java
Links links = Links.NONE;
Link link = Link.of("/foo");
		
List<NamedLinks> sources = Arrays.asList(
		NamedLinks.of("via varargs", links.and(link)),
		NamedLinks.of("via Iterable", links.and(Arrays.asList(link))),
		NamedLinks.of("via Stream", links.and(Stream.of(link))));

return DynamicTest.stream(sources.iterator(), NamedLinks::getName,
		it -> assertThat(it.links.getRequiredLink(IanaLinkRelations.SELF)).isNotNull());
```

The test is supposed to verify that *different parameter types* (here: some object, an `Iterable` of objects, and a stream) handed to the `Links.and(…)` method lead to a certain result. Ideally, I would like to treat those as sources but that would just hand in the common supertype to the `ThrowingConsumer`, so that I cannot call `links.and(…)` anymore as those expect the concrete types. To work around this limitation I call the method beforehand and end up with `Links` consistently but am then unable to produce test names in the second parameter as I cannot distinguish the individual `Links` instances anymore.

I created `NamedLinks` myself to be able to already eagerly bind a name to use in the setup of `Dynamic.stream(…)` but I wonder if that could be made available by JUnit OOTB providing a simple value type `NamedSource<T>` that exposes a name and source `T`and an overload of `DynamicTest.stream(Stream<NamedSource<T>> sources, ThrowingCallable<T> callback)`.

### PR DESCRIPTION
**Pull request by juliette-derancourt**
## Overview

Introduction of `DynamicTest.stream(Iterator<? extends Named<T>>, ThrowingConsumer<? super T>)` and `DynamicTest.stream(Stream<? extends Named<T>>, ThrowingConsumer<? super T>)`. Both methods use each name-value pair provided by the first parameter as the display name and value for each generated dynamic test.

Resolves #2375.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Review by marcphilipp — COMMENTED** [review]
LGTM 👍

**Comment by sbrannen** [comment]
Code and tests LGTM, too. 👍 

Just needs the Javadoc and release notes entry.

**Comment by juliette-derancourt** [comment]
> Just needs the Javadoc and release notes entry.

My favorite part! 😆

**Comment by sbrannen** [comment]
> > Just needs the Javadoc and release notes entry.
> 
> My favorite part! 😆

Now, now, let's not tell any tall stories. 😛

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc [hunk 1]
```diff
@@ -86,6 +86,9 @@ on GitHub.
 * In parameterized tests with `@MethodSource` or `@ArgumentSource`, arguments can now have
   optional names. When the argument is included in the display name of an iteration, this
   name will be used instead of the value.
+* `DynamicTests.stream()` can now consume named inputs, and will use those names as test
+  names.
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc:90]:
```suggestion
* `DynamicTests.stream()` can now consume `Named` input and will use each name-value
   pair as the display name and value for each generated dynamic test.
```

Comment (sbrannen) [reply documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc:90]: Probably good to link either to the Javadoc for `Named` or to a section in the User Guide covering the topic.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/DynamicTest.java [hunk 2]
```diff
@@ -155,6 +155,77 @@ public static DynamicTest dynamicTest(String displayName, URI testSourceUri, Exe
...
+	/**
+	 * Generate a stream of dynamic tests based on the given generator and test
+	 * executor.
+	 *
+	 * <p>Use this method when the set of dynamic tests is nondeterministic in
+	 * nature or when the input comes from an existing {@link Iterator}. See
+	 * {@link #stream(Stream, ThrowingConsumer)} as an alternative.
+	 *
+	 * <p>The given {@code inputGenerator} is responsible for generating
+	 * input values and test names. A {@link DynamicTest} will be added to
```
Comment (sbrannen) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/DynamicTest.java:167]:
```suggestion
	 * input values and display names. A {@link DynamicTest} will be added to
```

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/DynamicTest.java [hunk 3]
```diff
@@ -155,6 +155,77 @@ public static DynamicTest dynamicTest(String displayName, URI testSourceUri, Exe
...
+	/**
+	 * Generate a stream of dynamic tests based on the given input stream and
+	 * test executor.
+	 *
+	 * <p>Use this method when the set of dynamic tests is nondeterministic in
+	 * nature or when the input comes from an existing {@link Stream}. See
+	 * {@link #stream(Iterator, ThrowingConsumer)} as an alternative.
+	 *
+	 * <p>The given {@code inputStream} is responsible for supplying input values
+	 * and test names. A {@link DynamicTest} will be added to the resulting stream for
```
Comment (sbrannen) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/DynamicTest.java:202]:
```suggestion
	 * and display names. A {@link DynamicTest} will be added to the resulting stream for
```

**Review by sbrannen — CHANGES_REQUESTED** [review]

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc (hunk 1 above)
Comment (juliette-derancourt) [reply documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc]: You mean link as in clickable link ?

Comment (juliette-derancourt) [reply documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc]: Or just a mention?

Comment (sbrannen) [reply documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc]: Yes, I mean a hyperlink.

Comment (sbrannen) [reply documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc]: You added new examples in the `documentation` project. So you should be able to link to the section of the User Guide that displays those examples.

Comment (juliette-derancourt) [reply documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc]: But do we usually put links in the release notes?

Comment (sbrannen) [reply documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc]: Yes, when a new feature is introduced we often link to the corresponding section of the User Guide if such a section exists.

Typically it takes the form:

* Introduced feature XYZ
  * See `<link>` in the User Guide for details.

Comment (sbrannen) [reply documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc]: Although it appears that we have not provided such links in recent releases, here's an example where we did: https://junit.org/junit5/docs/5.5.0/release-notes/index.html#new-features-and-improvements-3

> Support for declarative timeouts using `@Timeout` or configuration parameters (see **User Guide** for details).

Comment (juliette-derancourt) [reply documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc]: Not sure I did it the right asciidoc-way 😄

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc [hunk 4]
```diff
@@ -86,6 +86,10 @@ on GitHub.
 * In parameterized tests with `@MethodSource` or `@ArgumentSource`, arguments can now have
   optional names. When the argument is included in the display name of an iteration, this
   name will be used instead of the value.
+* `DynamicTests.stream()` can now consume `Named` input and will use each name-value
+  pair as the display name and value for each generated dynamic test (see
+  link:{userguide-root}++/#writing-tests-dynamic-tests-examples++[User Guide] for details).
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc:91]:
```suggestion
  <<../user-guide/index.adoc#writing-tests-dynamic-tests-examples,User Guide>> for details).
```

File: documentation/src/docs/asciidoc/link-attributes.adoc [hunk 5]
```diff
@@ -2,6 +2,7 @@
 ifdef::backend-pdf[]
 :javadoc-root:                               https://junit.org/junit5/docs/{docs-version}/api
 endif::[]
+:userguide-root:                             https://junit.org/junit5/docs/{docs-version}/user-guide
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/link-attributes.adoc:5]:
```suggestion
```

**Review by marcphilipp — APPROVED** [review]

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc (hunk 4 above)
Comment (sbrannen) [reply documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc:91]: @marcphilipp beat me to it: I was just about to paste in `(see
  <<../user-guide/index.adoc#writing-tests-declarative-timeouts, User Guide>> for details)` as an example. ;-)

Comment (juliette-derancourt) [reply documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc:91]: Ah yes, the link is targeting the source file directly, got it ^^

Thanks @marcphilipp, and runner-up @sbrannen 😁

**Pull request merged**